google-cloud-spanner/lib/google/cloud/spanner/pool.rb

```diff
@@ -59,31 +61,40 @@ def with_session
         end
 
         def checkout_session
+          action = nil
           @mutex.synchronize do
             loop do
+              fail ClientClosedError if @closed
+
               read_session = session_queue.shift
               return read_session if read_session
               write_transaction = transaction_queue.shift
               return write_transaction.session if write_transaction
 
-              return new_session! if can_allocate_more_sessions?
+              if can_allocate_more_sessions?
+                action = :new
+                break
+              end
 
               fail SessionLimitError if @fail
 
               @resource.wait @mutex
             end
           end
+
+          return new_session! if action == :new
         end
 
         def checkin_session session
           unless all_sessions.include? session
             fail ArgumentError, "Cannot checkin session"
           end
 
+          session.reload!
```

Comment:
this will make rpcs, right? Why do we need to do this everytime a session is checked in?